src/analysis/souffle/tag_transforms.dl

```diff
@@ -0,0 +1,156 @@
+  [ policyRuleName: symbol,
+    result: SqlPolicyRuleResult,
+    preconditions: SqlPolicyRulePreconditionList ]
+
+.decl isSqlPolicyRule(rule: SqlPolicyRule)
+
+.decl sqlPolicyRuleNameHasPreconditionList(
+  policyRuleName: symbol, policyRuleListSuffix: SqlPolicyRulePreconditionList)
+
+sqlPolicyRuleNameHasPreconditionList(name, preconditionList) :-
+  isSqlPolicyRule([name, _, preconditionList]).
+
+.decl sqlPolicyRuleNameHasPreconditionListSuffix(
+  policyRuleName: symbol, policyRuleListSuffix: SqlPolicyRulePreconditionList)
+
+sqlPolicyRuleNameHasPreconditionListSuffix(name, preconditionList) :-
+  sqlPolicyRuleNameHasPreconditionList(name, preconditionList).
+
+sqlPolicyRuleNameHasPreconditionListSuffix(name, preconditionListTail) :-
+  sqlPolicyRuleNameHasPreconditionListSuffix(name, [_, preconditionListTail]).
+
+.decl policyRulePreconditionListLength(
+  policyRuleList: SqlPolicyRulePreconditionList, length: number)
+
+policyRulePreconditionListLength(nil, 0).
+policyRulePreconditionListLength([head, tail], tailLen + 1) :-
+  sqlPolicyRuleNameHasPreconditionListSuffix(_, [head, tail]),
+  policyRulePreconditionListLength(tail, tailLen).
+
+.decl sqlPolicyRuleNameHasPreconditionAtIndex(
+  policyRuleName: symbol, policyRulePrecondition: SqlPolicyRulePrecondition, index: number) 
+
+sqlPolicyRuleNameHasPreconditionAtIndex(name, precondition, suffixLen - 1) :-
+  sqlPolicyRuleNameHasPreconditionListSuffix(name, [precondition, tail]),
+  policyRulePreconditionListLength([precondition, tail], suffixLen).
+
+.decl sqlPolicyRuleNameHasResult(policyRuleName: symbol, result: SqlPolicyRuleResult)
+
+sqlPolicyRuleNameHasResult(name, result) :- isSqlPolicyRule([name, result, _]).
+
+isTag(tag) :- sqlPolicyRuleNameHasResult(_, $AddConfidentialityTag(tag)).
+isTag(tag) :- sqlPolicyRuleNameHasResult(_, $RemoveConfidentialityTag(tag)).
+isIntegrityTag(tag) :- sqlPolicyRuleNameHasResult(_, $AddIntegrityTag(tag)).
+
+.decl isTagTransformOperation(tagTransform: Operation)
+
+isTagTransformOperation(
+  [DEFAULT_SQL_OWNER, TAG_TRANSFORM_OPERATOR, result, operandList, attributes]) :-
+  isOperation([DEFAULT_SQL_OWNER, TAG_TRANSFORM_OPERATOR, result, operandList, attributes]).
+
+.type TagPrecondition = [ path: AccessPath, itag: IntegrityTag ]
+
+.decl tagTransformHasRuleName(tagTransform: Operation, ruleName: symbol)
+
+tagTransformHasRuleName(op, ruleName) :-
+  isTagTransformOperation(op),
+  operationHasAttribute(op, [TAG_TRANSFORM_RULE_NAME_ATTR, $StringAttributePayload(ruleName)]).
+
+.decl tagTransformHasResultAccessPath(tagTransform: Operation, resultPath: AccessPath)
+tagTransformHasResultAccessPath([owner, operator, result, operandList, attrs], result) :-
+  isTagTransformOperation([owner, operator, result, operandList, attrs]).
+
+.decl tagTransformHasIndexedTagPrecondition(
+  tagTransform: Operation, index: number, tagPrecondition: TagPrecondition)
+
+tagTransformHasIndexedTagPrecondition(op, index, [preconditionOperand, requiredTag]) :-
```

Comment:
is `index` needed at all? Don't we just care about `preconditionOperand`?

Reply:
Ah, the index is mainly needed to implement `forall` as used in `allTagTransformPreconditionsLessThanIndexMet`? Perhaps, document the reason somewhere?

Reply:
Added a comment explaining this.